### DISCUSSION TITLE
Drop MyISAM in SearchIndex table

### DISCUSSION THREAD

**Discussion by dolphiq-allard** [craftcms/cms#11374] [Ideas] [answered]
This question has been asked several times before, but let me bring this under attention one more time.

Would it be possible to drop the MyISAM engine in the `searchindex` table. We host several sites on DigitalOcean and they don't support it anymore, resulting in an error during installation. And since I always forget to remove this part before running the installation, I need to clean out the DB again before getting everything up and running.

Would be much appreciated.

Current situation:
```
// Add the FULLTEXT index on searchindex.keywords
$this->createTable(Table::SEARCHINDEX, [
    'elementId' => $this->integer()->notNull(),
    'attribute' => $this->string(25)->notNull(),
    'fieldId' => $this->integer()->notNull(),
    'siteId' => $this->integer()->notNull(),
    'keywords' => $this->text()->notNull(),
], ' ENGINE=MyISAM');
```

Preferable situation:
```
// Add the FULLTEXT index on searchindex.keywords
$this->createTable(Table::SEARCHINDEX, [
    'elementId' => $this->integer()->notNull(),
    'attribute' => $this->string(25)->notNull(),
    'fieldId' => $this->integer()->notNull(),
    'siteId' => $this->integer()->notNull(),
    'keywords' => $this->text()->notNull(),
]);
```

Thanks much

**Comment by brandonkelly** [comment, marked as answer]
Looks like fulltext indexes have been [available](https://dev.mysql.com/doc/refman/5.6/en/innodb-fulltext-index.html) for InnoDB tables since MySQL 5.6.4, and since Craft 4 requires MySQL 5.7.8+, I don’t think there’s any good reason not to do this. So we’ll look into it for 4.1.

**Reply by dolphiq-allard** [reply]
Thanks for the quick reply (as always). Looking forward to this small improvement.

**Reply by angrybrad** [reply]
Change made for 4.1! https://github.com/craftcms/cms/commit/3878470ec45bec2a906924dd696897dc26eee84f

**Reply by dolphiq-allard** [reply]
Thanks all, looking forward.

**Reply by brandonkelly** [reply]
Craft 4.1 is out with this change ✨

**Reply by dolphiq-allard** [reply]
Super nice, looking forward to this and all the other additions in this version. Keep up the good work, you guys created a stunning CMS!

**Reply by jonaswouters** [reply]
This still doesn't work out-of-the-box for DigitalOcean, unfortunately. 

By altering the Install.php file, I made it work. I think the result is the same with less LoC.

```php
        if ($this->db->getIsMysql()) {
            $this->createIndex(null, Table::ELEMENTS_SITES, ['uri', 'siteId']);
            $this->createIndex(null, Table::USERS, ['email']);
            $this->createIndex(null, Table::USERS, ['username']);

            // Add the FULLTEXT index on searchindex.keywords
            $this->createTable(Table::SEARCHINDEX, [
                'elementId' => $this->integer()->notNull(),
                'attribute' => $this->string(25)->notNull(),
                'fieldId' => $this->integer()->notNull(),
                'siteId' => $this->integer()->notNull(),
                'keywords' => $this->text()->notNull(),
'PRIMARY KEY([[elementId]], [[attribute]], [[fieldId]], [[siteId]])',
            ]);

            $sql = 'ALTER TABLE ' .
                $this->db->quoteTableName(Table::SEARCHINDEX) . ' ' .
                'ADD FULLTEXT (' . $this->db->quoteColumnName('keywords') . ')';

            $this->db->createCommand($sql)->execute();
        } else {
```

**Reply by brandonkelly** [reply]
@jonaswouters `CREATE FULLTEXT INDEX` and `ALTER TABLE … ADD FULLTEXT` [appear](https://stackoverflow.com/a/35586145/1688568) to just be aliases of each other. Did you get a specific error with the former?

**Reply by jonaswouters** [reply]
> @jonaswouters `CREATE FULLTEXT INDEX` and `ALTER TABLE … ADD FULLTEXT` [appear](https://stackoverflow.com/a/35586145/1688568) to just be aliases of each other. Did you get a specific error with the former?

Sorry for the late reply. This is the error: 
I just did another install, same error, same solution. 

```
*** failed to install Craft: An error occurred while executing the "craft\migrations\Install migration: SQLSTATE[HY000]: General error: 3750 Unable to create or change a table without a primary key, when the system variable 'sql_require_primary_key' is set. Add a primary key to the table or unset this variable to avoid this message. Note that tables without a primary key can cause performance problems in row-based replication, so please consult your DBA before changing this setting.
The SQL being executed was: CREATE TABLE `searchindex` (
	`elementId` int(11) NOT NULL,
	`attribute` varchar(25) NOT NULL,
	`fieldId` int(11) NOT NULL,
	`siteId` int(11) NOT NULL,
	`keywords` text NOT NULL
) ENGINE = InnoDb DEFAULT CHARACTER SET = utf8
```

**Reply by brandonkelly** [reply]
@jonaswouters Ahh… I missed that you had moved the primary key definition to the `CREATE TABLE` statement. The error message is complaining about the `CREATE TABLE` command, so that must be the important change there, not the switch from `CREATE FULLTEXT INDEX` to `ALTER TABLE … ADD FULLTEXT`.

Made the same change for the next Craft 3 and 4 releases (e2baa64a3e1969b0fdb6c32bf5a723276fe5142b) so this should be fully resolved going forward!

**Reply by brandonkelly** [reply]
Craft 3.7.55 and 4.2.5 are out with that change.
